### DISCUSSION TITLE
Rescheduled Activity Functions Concurrency Concern

### DISCUSSION THREAD

**Discussion by epDugas** [Azure/azure-functions-durable-extension#3002] [Q&A] [answered]
An activity function's execution is guaranteed to be "at-least-once". Because the activity function can be retried for a myriad of reasons, it is recommended to make the function idempotent. I am looking for clarification on if it is possible that a subsequent execution can occur **concurrently** with a previous execution. 

Scenario:
- Activity Func 1 is scheduled, it's work item is pulled from the queue, and execution begins (E1).
- The work item becomes visible again, and is pulled from the queue, and a subsequent execution begins (E2).

Is it possible that E1 and E2 could execute concurrently in any way?

Thanks.

**Comment by cgillum** [comment, marked as answer]
It's theoretically possible but would be extremely rare. The case where I think it's possible is:

1. Worker 1 dequeues the message but somehow gets stuck trying to process it (pre-E1).
2. Message times out and becomes visible again
3. Worker 2 dequeues the same message, initiating E2 immediately
4. Worker 1 gets unstuck, and E1 actually starts.

So in this case, the message would have been dequeued at very different times (e.g., 5 minutes apart from each other) but the actual executions start at the same time (or overlap in some way).

**Reply by epDugas** [reply]
Thanks for the clarification.
